doc/reference/trd-storage-permissions.md

```diff
@@ -0,0 +1,174 @@
+Application Persistent Data Storage Permissions
+===============================================
+
+**TRD:** <br/>
+**Working Group:** Kernel<br/>
+**Type:** Documentary<br/>
+**Status:** Draft <br/>
+**Author:** Brad Campbell<br/>
+**Draft-Created:** 2024/06/06<br/>
+**Draft-Modified:** 2024/06/06<br/>
+**Draft-Version:** 1<br/>
+**Draft-Discuss:** devel@lists.tockos.org<br/>
+
+
+Abstract
+-------------------------------
+
+Tock supports storing persistent state for applications, and all persistent
+state in Tock is identified based on the application that stored it. Tock
+supports permissions for persistent state, allowing for the kernel to restrict
+which applications can store state and which applications can read stored state.
+This TRD describes the permissions architecture for persistent state in Tock.
+This document is in full compliance with [TRD1][TRD1].
+
+
+1 Introduction
+-------------------------------
+
+Tock applications need to be able to store persistent state. Additionally,
+applications need to be able to keep data private from other applications. The
+kernel should also be able to allow specific applications to read and modify
+state from other applications.
+
+This requires a method for assigning applications persistent identifiers,
+a mechanism for granting storage permissions to specific applications,
+and kernel abstractions for implementing storage capsules that respect
+the storage permissions.
+
+
+2 Scope
+-------------------------------
+
+This document only describes the permission architecture in Tock for supporting
+application persistent storage. This document does not include specific types of
+persistent storage (e.g., flash, FRAM, etc.) or storage access abstractions
+(e.g., block-access, byte-access, etc.).
+
+
+3 Permissions
+-------------------------------
+
+All persistent application data is labeled based on the application which wrote
+the data. By default, applications can always read and modify data they wrote.
+Applications can read and modify data from other applications with suitable
+permissions.
+
+There are three types of permissions:
+
+1. **Write**: The application can write data.
+1. **Read**: The application can read other applications' data.
+1. **Modify**: The application can modify other applications' data.
+
+Each permission type is independent. For example, an application can be given
+read permission for specific data but not be able to write new data itself.
+
+
+4 Requirements
+-------------------------------
+
+The Tock storage model imposes the following requirements:
+
+1. Applications are given separate write, read, and modify permissions.
+2. The label stored with the persistent data when the data are written is the
+   application's short AppID.
+3. Applications without a `ShortId::Fixed` cannot access (i.e.,
+   read/write/modify) any persistent storage.
```

Comment:
Why have this limitation?

Reply:
This makes it clear how apps are identified for storage.

Reply:
Some more explanation in 3 on how IDs are chosen / why those are used might help explain this?

Reply:
There is a TRD (trd-appid.md) which explains short IDs.

Reply:
I don't follow. I'm not sure this helps make it clear

Reply:
Since there is only one way for Tock to assign storage IDs, Tock users/developers know exactly how they are determined in all cases.

Reply:
I think the problem with this, however, while there is "one way to assign storage IDs", i.e. they are short IDs, there is _not_ "one way to assign short IDs", thus by transitivity, there is pretty trivially no "one way to assign a storage ID"

Reply:
(Which, as discussed on core call / in summary note below, is okay!)